src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -0,0 +1,176 @@
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings
+{
+    internal abstract class AbstractRefactoringHelpersService : IRefactoringHelpersService
+    {
+        public async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+        {
+            return await TryGetSelectedNodeAsync(document, selection, n => n is TSyntaxNode, cancellationToken).ConfigureAwait(false) as TSyntaxNode;
+        }
+
+        public Task<SyntaxNode> TryGetSelectedNodeAsync(Document document, TextSpan selection, Predicate<SyntaxNode> predicate, CancellationToken cancellationToken)
+        {
+            return TryGetSelectedNodeAsync(document, selection, predicate, DefaultNodeExtractor, cancellationToken);
+        }
+
+        public async Task<SyntaxNode> TryGetSelectedNodeAsync(Document document, TextSpan selection, Predicate<SyntaxNode> predicate, Func<SyntaxNode, ISyntaxFactsService, SyntaxNode> extractNode, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var selectionStripped = await CodeRefactoringHelpers.GetStrippedTextSpan(document, selection, cancellationToken).ConfigureAwait(false);
+
+            // Handle selections
+            // - the smallest node whose span inlcudes whole selection
+            // - extraction from such node is attempted, result tested via predicate
+            // - travels upwards through same-sized nodes, extracting and testing predicate
+            var node = root.FindNode(selectionStripped, getInnermostNodeForTie: true);
+            SyntaxNode prevNode;
+            do
+            {
+                var wantedNode = TryGetAcceptedNodeOrExtracted(node, predicate, extractNode, syntaxFacts);
+                if (wantedNode != default)
+                {
+                    return wantedNode;
+                }
+
+                prevNode = node;
+                node = node.Parent;
+            }
+            while (node != null && prevNode.FullWidth() == node.FullWidth());
+
+            // consider what selection is touching only when it's empty -> prevents
+            // `[|C|] methodName(){}` from registering as relevant for method Node
+            if (!selection.IsEmpty)
+            {
+                return default;
+            }
+
+            // get Token for current selection (empty) location
+            var tokenOnSelection = root.FindToken(selectionStripped.Start);
+
+            // Token to right or containing current selection
+            var tokenToRightOrIn = tokenOnSelection.Span.Contains(selectionStripped.Start)
+                ? tokenOnSelection
+                : default;
+
+            if (tokenToRightOrIn != default)
```

Comment:
i think it would be valuable to have a comment explaining what this entire block is trying to accomplish at a conceptual level.